### PR TITLE
BestBQ implementation

### PR DESCRIPTION
**Pull request by Judyxujj**
add the two components, i.e. mask and quantiser, for BestRQ.

### CODE REVIEW

File: i6_models/parts/best_rq/quantizer.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        # projection matrix use Xavier initialization
+        P_init = torch.empty((input_dim, codebook_dim))
+        self.register_buffer("P", nn.init.xavier_uniform_(P_init))
+
+        # normalize random matrix for codebook
+        self.register_buffer("CB", F.normalize(torch.randn(codebook_num_vars, codebook_dim)))
+
+    def forward(self, x: torch.tensor) -> torch.tensor:
+        x = F.normalize(x @ self.P)
+        return vector_norm((self.CB.unsqueeze(1) - x.unsqueeze(1)), dim=-1).argmin(dim=1)
```
Comment (michelwi) [inline i6_models/parts/best_rq/quantizer.py:37]: I am wondering about the dimensions here:

`x` starts out as `[B, T, input_dim]` and is then transformed to `[B, T, codebook_dim]` (?)
`CB` has `[codebook_num_vars, codebook_dim]`.

Then should the unsqueezing not be like `self.CB.unsqueeze(0).unsqueeze(0)` to add `B,T` and `x.unsqueeze(1)` to add `codebook_num_vars` so that both have `[B, T, codebook_num_vars, codebook_dim]`.

And then we would do `vector_norm([...], dim=-1)` to get `[B, T, codebook_num_vars]` (with the values being the respective norms) 
followed by `[...].argmin(dim=-1)` which yields `[B, T]` (with the values being the index in the codebook `CB` that best matches the projected `x`.

And then here the indices of the best match are returned.. is this intended? Without the knowledge of the codebook the indices seem worthless to me. Should the returned value not rather be value of the codebook at the specified index?
I should read the paper...
[a few minutes later]
I read the paper: the approach indeed uses the indices of the codebook as targets for a classifier. 
So this part is correct.

**Review by michelwi — CHANGES_REQUESTED** [review]
Not sure if there are problems in the code or if I misunderstood what it is doing 🤷

**Comment by albertz** [comment]
Does it make sense to include this in i6_models? What exactly is our policy w.r.t. what we want to have here? I thought i6_models is intended for models (and maybe other functions) which are supposed to be used by a larger amount of people among us. But it is not intended for research code, where you want to try something out. Here, as I understand, this is mostly adapted from Fairseq, so maybe it's already well tested, but I don't know. Did you test this?

File: i6_models/parts/best_rq/mask.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        mask_replace_val: str,
+        mask_prob: float,
+        mask_length: int,
+        min_masks: int = 0,
+    ):
+        """
+        :param input_dim: number of feature dimension of input
+        :param mask_replace_val: the way to replace masked frames, either with zeros or lernable embeddings
+        :param mask_prob: percentage of frames to be masked out
+        :param mask_length: the length of each mask span
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply i6_models/parts/best_rq/mask.py:30]: Here the two masking regions could overlap and form a longer region,  do you think such issues should be avoided?  In Specaug I think we would also have such masking regions overlap issue, I am not sure whether it matter that much

File: i6_models/parts/best_rq/mask.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        input_dim: int,
+        mask_replace_val: str,
+        mask_prob: float,
+        mask_length: int,
+        min_masks: int = 0,
+    ):
+        """
+        :param input_dim: number of feature dimension of input
+        :param mask_replace_val: the way to replace masked frames, either with zeros or lernable embeddings
+        :param mask_prob: percentage of frames to be masked out
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply i6_models/parts/best_rq/mask.py]: It should be percentage so I rename the parameter

File: i6_models/parts/best_rq/mask.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            num_mask = int(
+                # add a random number for probabilistic rounding
+                self.mask_prob * seq_len / float(self.mask_length)
+                + np.random.rand()
+            )
+            num_mask = max(self.min_masks, num_mask)
+
+            min_len = self.mask_length
+            if seq_len - min_len <= num_mask:
+                min_len = seq_len - num_mask - 1
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply i6_models/parts/best_rq/mask.py:66]: I remove the `self.min_masks` parameter

File: i6_models/parts/best_rq/mask.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            mask_idc = np.random.choice(seq_len - min_len, num_mask, replace=False)
```
[the remark this answers is not in the document]
Comment (Judyxujj) [reply i6_models/parts/best_rq/mask.py:67]: I suspect that it would not matter that much from our previous experience with Specaug. In our current Specaug implementation, the mask beginning indices are generated like this
```
    z = torch.rand((batch_dim_size, tensor.shape[axis]), device=tensor.device)  # [B,dim]
    _, indices = torch.topk(z, max_num_masks, dim=1)
```
it is also not guaranteed that two masks do not overlap, but it does not seem to be an issue, but I think we never compared with the version where overlap is avoided.

**Comment by Judyxujj** [comment]
> Does it make sense to include this in i6_models? What exactly is our policy w.r.t. what we want to have here? I thought i6_models is intended for models (and maybe other functions) which are supposed to be used by a larger amount of people among us. But it is not intended for research code, where you want to try something out. Here, as I understand, this is mostly adapted from Fairseq, so maybe it's already well tested, but I don't know. Did you test this?

I think BestRQ is a very general pre-training architecture that we might use later also in the group, so I add it here. I ran them with a small dataset and seems to work.

**Review by michelwi — APPROVED** [review]

**Comment by curufinwe** [comment]
> Does it make sense to include this in i6_models? What exactly is our policy w.r.t. what we want to have here? I thought i6_models is intended for models (and maybe other functions) which are supposed to be used by a larger amount of people among us. But it is not intended for research code, where you want to try something out. Here, as I understand, this is mostly adapted from Fairseq, so maybe it's already well tested, but I don't know. Did you test this?

I have no problem adding also some code that is a bit more "research-y" to the repo. But of course once it's in we would require that new variants get a separate class name to not break the old behavior (bugs are the exception of course). If we see that some model / part of a model is changing rapidly we might want to delay things until it settles down a bit, but in this instance I would expect the code from fairseq to broadly work.

File: i6_models/parts/best_rq/mask.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+            num_mask = int(
+                # add a random number for probabilistic rounding
+                self.mask_percentage * seq_len / float(self.mask_length)
+                + np.random.rand()
+            )
+
+            min_len = self.mask_length
+            if seq_len - min_len <= num_mask:
+                min_len = seq_len - num_mask - 1
```
Comment (kuacakuaca) [inline i6_models/parts/best_rq/mask.py:66]: this seems to be incorrect. Let's say seq_len = 5, mask_percentage = 1.0, mask_length = 5, then num_mask could be 1. In this case, seq_len - mask_length = 0 <= 1, then min_len is assigned 5 - 1 -1 = 3. Through code in next line, np.random.choice(2, 1, replace=False) makes it possible that not 100% frames are masked out. Pls correct me if I'm wrong.

**Review by kuacakuaca — CHANGES_REQUESTED** [review]

File: i6_models/parts/best_rq/mask.py (hunk 6 above)
Comment (michelwi) [reply i6_models/parts/best_rq/mask.py:66]: I can follow your logic and think that
```
if seq_len - min_len < num_mask:
     min_len = seq_len - num_mask
```
would be more true to the intended implementation.

But this case will most likely not appear as nobody would want to mask the entire sequence.

**Review by kuacakuaca — APPROVED** [review]
